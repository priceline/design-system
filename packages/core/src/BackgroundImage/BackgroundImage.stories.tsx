import type { Meta } from '@storybook/react'
import { Box } from '../Box/Box'
import { Card } from '../Card/Card'
import { Text } from '../Text/Text'
import { BackgroundImage } from './BackgroundImage'
import { argTypes, defaultArgs } from './BackgroundImage.stories.args'

export default {
  title: 'core / BackgroundImage',
  component: BackgroundImage,
  args: defaultArgs,
  argTypes,
  decorators: [
    (story) => (
      <Box height='600px' width='100%'>
        {story()}
      </Box>
    ),
  ],
  parameters: {
    docs: {
      description: {
        component: `Use the <code>&lt;BackgroundImage /&gt;</code> component to render a
    background image. Use the *variation* prop to change the attachment mode
    of the background.`,
      },
    },
  },
} as Meta

const Template = (args) => <BackgroundImage {...args} />

export const _BackgroundImage = Template.bind({})

const StoryChildren = () => (
  <Box p={4}>
    <Text fontSize={6} fontWeight='bold' textAlign='center' color='white'>
      Hello
    </Text>
  </Box>
)

export const FixedHeight = Template.bind({})
FixedHeight.args = {
  height: '320px',
  width: '360px',
  children: <StoryChildren />,
}

export const FixedIntHeight = Template.bind({})
FixedIntHeight.storyName = 'Fixed Height (Integer)'
FixedIntHeight.args = {
  height: 400,
  width: '360px',
  children: <StoryChildren />,
}

export const Responsive = Template.bind({})
Responsive.args = {
  height: ['100px', '300px', '600px'],
  width: ['100%', '216px', '260px'],
  children: <StoryChildren />,
}
Responsive.parameters = {
  viewport: { defaultViewport: 'designSystem_xs' },
}

const RoundedTemplate = (args) => (
  <Box width='300px'>
    <BackgroundImage {...args} />
    <Card rounded='bottom' height='100px'>
      <Text textAlign='center' my={3}>
        Hello Priceline!
      </Text>
    </Card>
  </Box>
)

export const Rounded = RoundedTemplate.bind({})
Rounded.args = {
  borderRadius: 'full',
  rounded: 'top',
  height: '100px',
}

export const BackgroundPosition = Template.bind({})
BackgroundPosition.args = {
  height: 400,
  width: '360px',
  children: <StoryChildren />,
  backgroundPosition: 'left',
}
