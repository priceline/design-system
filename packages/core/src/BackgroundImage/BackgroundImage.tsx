import React, { type ComponentPropsWithoutRef } from 'react'
import styled, { css } from 'styled-components'
import {
  BorderRadiusProps,
  HeightProps,
  WidthProps,
  borderRadius,
  compose,
  height,
  width,
} from 'styled-system'
import { applyVariations, getPaletteColor } from '../utils'
import { borderRadiusAttrs } from '../utils/attrs/borderRadiusAttrs'

const variations = {
  parallax: css`
    background-attachment: fixed;
  `,
  static: css`
    background-attachment: scroll;
  `,
}

/**
 * @public
 */
export const backgroundPositionList = ['top', 'bottom', 'left', 'right', 'center']
const image = (props) => (props.image ? { backgroundImage: `url(${props.image})` } : null)

/**
 * @public
 */
export type BackgroundImageProps = WidthProps &
  HeightProps &
  BorderRadiusProps &
  Omit<ComponentPropsWithoutRef<'img'>, 'width' | 'height'> & {
    variation?: 'parallax' | 'static'
    image?: string
    borderRadius?: string
    rounded?: string
    backgroundPosition?: (typeof backgroundPositionList)[number]
  }

/**
 * @public
 */
export const BackgroundImage: React.FC<BackgroundImageProps> = styled.div.attrs(borderRadiusAttrs)`
  background-position: ${(props) => props.backgroundPosition};
  background-size: cover;
  background-repeat: no-repeat;
  background-color: ${getPaletteColor('border.light')};
  ${applyVariations('BackgroundImage', variations)}
  ${image} 
  ${(props) => compose(height, width, borderRadius)(props)}
`

BackgroundImage.defaultProps = {
  variation: 'static',
  backgroundPosition: 'center',
}

BackgroundImage.displayName = 'BackgroundImage'
