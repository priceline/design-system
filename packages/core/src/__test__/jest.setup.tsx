import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
