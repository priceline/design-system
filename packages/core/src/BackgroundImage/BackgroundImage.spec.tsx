import { render, screen } from '../__test__/testing-library'
import { BackgroundImage } from './BackgroundImage'

describe('BackgroundImage', () => {
  test('renders', () => {
    const { asFragment } = render(<BackgroundImage />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders with image', () => {
    const { asFragment } = render(<BackgroundImage image='hello.png' />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders with height', () => {
    const { asFragment } = render(<BackgroundImage height='320px' />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders with border radius', () => {
    const { asFragment } = render(
      <BackgroundImage height='320px' borderRadius='full' rounded='top' data-testid='background-image' />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders with background position', () => {
    render(
      <BackgroundImage
        height='320px'
        borderRadius='full'
        rounded='top'
        backgroundPosition='top'
        data-testid='background-image'
      />
    )

    const element = screen.getByTestId('background-image')
    expect(element).toHaveStyleRule('background-position', 'top')
  })

  test('renders with default background position', () => {
    render(
      <BackgroundImage height='320px' borderRadius='full' rounded='top' data-testid='background-image' />
    )
    const element = screen.getByTestId('background-image')
    expect(element).toHaveStyleRule('background-position', 'center')
  })

  describe('variations', () => {
    describe('parallax variation', () => {
      test('default render', () => {
        render(<BackgroundImage variation='parallax' image='hello.png' data-testid='background-image' />)
        const element = screen.getByTestId('background-image')
        expect(element).toHaveStyleRule('background-attachment', 'fixed')
      })
    })
  })
})
