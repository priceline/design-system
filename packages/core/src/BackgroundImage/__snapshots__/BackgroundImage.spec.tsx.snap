// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackgroundImage renders 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-position: center;
  background-size: cover;
  background-repeat: no-repeat;
  background-color: #eceff2;
  background-attachment: scroll;
}

<div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</DocumentFragment>
`;

exports[`BackgroundImage renders with border radius 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-position: center;
  background-size: cover;
  background-repeat: no-repeat;
  background-color: #eceff2;
  background-attachment: scroll;
  height: 320px;
  border-radius: 9999px 9999px 0 0;
}

<div
    class="c0"
  >
    <div
      class="c1"
      data-testid="background-image"
      height="320px"
    />
  </div>
</DocumentFragment>
`;

exports[`BackgroundImage renders with height 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-position: center;
  background-size: cover;
  background-repeat: no-repeat;
  background-color: #eceff2;
  background-attachment: scroll;
  height: 320px;
}

<div
    class="c0"
  >
    <div
      class="c1"
      height="320px"
    />
  </div>
</DocumentFragment>
`;

exports[`BackgroundImage renders with image 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c1 {
  background-position: center;
  background-size: cover;
  background-repeat: no-repeat;
  background-color: #eceff2;
  background-attachment: scroll;
  background-image: url(hello.png);
}

<div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</DocumentFragment>
`;
